### PR TITLE
Stein mixture 

### PR DESCRIPTION
**Pull request by OlaRonning**
In merging the original Stein mixture paper with ours, we found an inconsistency in the original article. We correct that here and can (probably) connect the current form with optimizing the ELBO (hence a lower bound on the marginal likelihood). Feel free to PM me for details if you are interested.

It was related to #1590. Rebase failed.

## TODO
- [x] Update documentation
- [x] Update tests.
- [x] Refactor/rework [stein loss](https://github.com/aleatory-science/numpyro/blob/feature/sm/numpyro/contrib/einstein/stein_loss.py#L12)
- [x] Add test for `SteinLoss`
- [x] Refactor/rework [Predictive](https://github.com/aleatory-science/numpyro/blob/feature/sm/numpyro/infer/util.py#L1012) broadcasting solution for mixture guides.
- [x] Add test for `MixtureGuidePredictive`

### CODE REVIEW

**Comment by OlaRonning** [comment]
I will update the docs in a separate PR. This PR is already huge (sorry about that c:).

File: numpyro/contrib/einstein/mixture_guide_predictive.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+from jax import numpy as jnp, random, tree_map, vmap
+from jax.tree_util import tree_flatten
+
+from numpyro.handlers import substitute
+from numpyro.infer import Predictive
+from numpyro.infer.util import _predictive
+
+
+class MixtureGuidePredictive:
```
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/contrib/einstein/mixture_guide_predictive.py:15]: It is useful in some cases, e.g. to do predictive with (num_chains, num_samples) or for autoguide posterior samples with multiple sample shapes.

**Review by fehiepsi — APPROVED** [review]

**Comment by fehiepsi** [comment]
LGTM, thanks Ola!

File: numpyro/contrib/einstein/mixture_guide_predictive.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (OlaRonning) [reply numpyro/contrib/einstein/mixture_guide_predictive.py:15]: Hmm, the core is I need the mixture assignments to sample from the model. I specialized the ELBO to SteinLoss account for this. 

I suppose the guide is the shared element for both cases (loss and predictive sampling) so hiding the complications there seems more modular. I'll try to lift the guide to a mixture guide (and have it as a  SteinVI attribute)--then the lifted guide would work with current Predictive and ELBO. Could also simplify some logic in SteinVI. I'll sketch it and get back.
